ADPPRG-126821: Reduce contention in access to pdx registry

albertogpz opened the pull request:
Access to the Geode predefined types in the
serialization registry has been changed from a
mutex synchronized map to a hardwired switch-case
in order to avoid contention when several threads
access them.

albertogpz committed:
GEODE-10400: Fix metadata completion detection

 - Whenever execution server functions with isHA=false, hasResult=true
   and optimizeForWrite=true, if the metadata is incomplete a request of
   type EXECUTE_REGION_FUNCTION_SINGLE_HOP is sent with an invalid
   bucket set partition. This causes an
   InternalFunctionInvocationTargetException exception
 - In order to solve this issue, now, whenever calling
   groupByServerToBuckets, if there isn't a valid location for one of
   the buckets, it returns nullptr, triggering the fallback mechanism
   which would be sending EXECUTE_REGION_FUNCTION to one of the nodes
   instead.

albertogpz committed:
GEODE-10416: Bump to boost 1.80.0

albertogpz committed:
GEODE-10414: Implement Region::putIfAbsent method

 - Moved PUT actions from LocalRegion into its own modules.
 - Added PutIfAbsent action.
 - Added the necessary scaffolding for put requests to return a value.
 - Moved TcrMessagePut implementation into its own file. Now this message
   type accepts an operation as input.
 - Moved TcrMessageDestroy implementation into its own file. Now this
   message accepts an operation as input.
 - Also, updated TcrMessageDestroy operation format so it uses a byte
   part instead of an object.
 - Added putIfAbsent method to the Region interface and its
   implementations both for LocalRegion and ThinClientRegion.
 - Modified LocalRegion and ThinClientRegion to comply with the above
   code changes.
 - Modified PUT's old value parsing to handle null pointer.
 - Also replaced Cacheable type by Serializable in some places in order
   to take advantage of fast-forward definition.

ecealgp committed:
ADPPRG-126821: Reduce contention in access to pdx registry

Access to the Geode predefined types in the
serialization registry has been changed from a
mutex synchronized map to a hardwired switch-case
in order to avoid contention when several threads
access them.

gaussianrecurrence commented on cppcache/src/SerializationRegistry.cpp:
```diff
@@ -605,19 +591,162 @@ void TheTypeMap::findDataSerializable(int32_t id,
-    func = found->second;
+  // Geode builtins will be inside the switch value cases.
+  // The type ids should be in DSCode.hpp
+  switch (dsCode) {
```
You don't need to modify this function, since the only calling function, SerializationRegistry::serialize, is already handling library pre-defined types, and there is no other call point. Hence, this function will be only called if the type we are looking for is an user-defined type

albertogpz replied:
Good catch. Thanks for pointing it out.

gaussianrecurrence commented on cppcache/src/SerializationRegistry.cpp:
```diff
@@ -321,53 +239,121 @@ std::shared_ptr<Serializable> SerializationRegistry::deserialize(
-
-  theTypeMap_.findDataSerializablePrimitive(dsCode, createType);
+  if (createType == nullptr) {
+    theTypeMap_.findDataSerializablePrimitive(dsCode, createType);
```
It would be better if this is handled in the default case of above's switch, making it more clear this handles user-defined types and avoiding the need to make the null comparison

albertogpz replied:
I agree.

msdt93 reviewed (changes requested):
Also, about the commit message, the SerializationRegistry and TheTypeMap handles several different types, not just PdxTypes. PdxTypes are one of the types handled by these registries and in particular, the contention is not caused by PdxTypes.

ecealgp committed:
Changes after review.

Final commit must be modified as follows:
Access to the Geode predefined types in the
serialization registry has been changed from a
mutex synchronized map to a hardwired switch-case
in order to avoid contention when several threads
access them.

albertogpz commented on cppcache/src/SerializationRegistry.cpp:
```diff
@@ -605,19 +591,162 @@ void TheTypeMap::findDataSerializable(int32_t id,
-  const auto& found = dataSerializableFixedIdMap_.find(dsfid);
-  if (found != dataSerializableFixedIdMap_.end()) {
-    func = found->second;
+  switch (dsfid) {
```
Good point. I agree.

gaussianrecurrence commented on cppcache/src/SerializationRegistry.cpp:
```diff
@@ -605,20 +623,20 @@ void TheTypeMap::findDataSerializable(int32_t id,
-  if (found != dataSerializableFixedIdMap_.end()) {
-    func = found->second;
-  }
+    const std::lock_guard<std::mutex> guard(dataSerializableFixedIdMapMutex_);
```
Clang formatting rules establish that 2 spaces should be used for indentation, I think you are using 4 here.

gaussianrecurrence commented on cppcache/src/SerializationRegistry.cpp:
```diff
@@ -605,20 +623,20 @@ void TheTypeMap::findDataSerializable(int32_t id,
-  if (found != dataSerializablePrimitiveMap_.end()) {
-    func = found->second;
-  }
+    const std::lock_guard<std::mutex> guard(dataSerializablePrimitiveMapMutex_);
```
Clang formatting rules establish that 2 spaces should be used for indentation, I think you are using 4 here.

ecealgp committed:
Indentation changes after review.

msdt93 reviewed (changes requested):
Apparently, after changing release-nordix/1.15.1, this requires a rebase
